--- ../pgbouncer/include/bouncer.h	2020-04-12 09:52:43.000000000 +0300
+++ ./include/bouncer.h	2020-05-12 21:02:57.000000000 +0300
@@ -87,6 +87,12 @@
 
 extern int cf_sbuf_len;
 
+
+/* pgbouncer-rr extensions */
+#include "pycall.h"
+#include "route_connection.h"
+#include "rewrite_query.h"
+
 #include "util.h"
 #include "iobuf.h"
 #include "sbuf.h"
@@ -335,6 +341,7 @@
 
 	SocketState state:8;	/* this also specifies socket location */
 
+	long last_tx_timestamp:1; /* DBEELINE - client attribute for read after write race condition */
 	bool ready:1;		/* server: accepts new query */
 	bool idle_tx:1;		/* server: idling in tx */
 	bool close_needed:1;	/* server: this socket must be closed ASAP */
