--- ../pgbouncer/src/client.c	2020-04-12 12:53:20.000000000 +0300
+++ src/client.c	2020-04-12 12:16:46.000000000 +0300
@@ -695,6 +695,19 @@
 	if (client->pool->db->admin)
 		return admin_handle_client(client, pkt);
 
+	/* pgbouncer-rr extensions: query rewrite & client connection routing */
+	if(client->link && client->link->idle_tx) {
+		if(!find_server(client)) {
+			return false;
+		}
+	}
+	if (pkt->type == 'Q' || pkt->type == 'P') {
+		if (!rewrite_query(client, pkt)) {
+			return false;
+		}
+        route_client_connection(client, pkt);
+	}
+
 	/* acquire server */
 	if (!find_server(client))
 		return false;
