--- ../pgbouncer/src/client.c	2020-04-12 12:53:20.000000000 +0300
+++ ./src/client.c	2020-05-09 22:18:47.000000000 +0300
@@ -1,3 +1,4 @@
+
 /*
  * PgBouncer - Lightweight connection pooler for PostgreSQL.
  *
@@ -630,7 +631,6 @@
 {
 	SBuf *sbuf = &client->sbuf;
 	int rfq_delta = 0;
-
 	switch (pkt->type) {
 
 	/* one-packet queries */
@@ -695,6 +695,23 @@
 	if (client->pool->db->admin)
 		return admin_handle_client(client, pkt);
 
+	/* pgbouncer-rr extensions: query rewrite & client connection routing */
+	if(client->link) {
+		slog_info(client, "DBEELINE - client link is in state %s", client->link->state);
+	}
+	if(client->link && client->link->idle_tx) {
+		slog_info(client, "SKIPPING ROUTING RULES: client is transacting");
+		return skip_query_interception(client, pkt, sbuf, rfq_delta);
+	}
+	
+	if (pkt->type == 'Q' || pkt->type == 'P') {
+		slog_info(client, "Conditions for rr patch met");
+		if (!rewrite_query(client, pkt)) {
+			return false;
+		}
+        route_client_connection(client, pkt);
+	}
+
 	/* acquire server */
 	if (!find_server(client))
 		return false;
